drivers/usb/device/usb_dc_sam_usbc.c

```diff
@@ -0,0 +1,1160 @@
+		    uint32_t data_len, uint32_t *ret_bytes)
+{
+	uint8_t ep_idx = EP_ADDR2IDX(ep);
+	uint8_t ep_bank;
+	uint32_t packet_len;
+	uint32_t key;
+
+	if (ep_idx >= NUM_OF_EP_MAX) {
+		LOG_ERR("wrong endpoint index/address");
+		return -EINVAL;
+	}
+
+	if (!usb_dc_ep_is_enabled(ep_idx)) {
+		LOG_ERR("endpoint not enabled");
+		return -ENODEV;
+	}
+
+	if (EP_ADDR2DIR(ep) != USB_EP_DIR_IN) {
+		LOG_ERR("wrong endpoint direction");
+		return -EINVAL;
+	}
+
+	if ((regs->UECON[ep_idx] & USBC_UECON0_STALLRQ) != 0) {
+		LOG_WRN("endpoint is stalled");
+		return -EBUSY;
+	}
+
+	/* Check if there is bank available */
+	if (ep_idx > 0) {
+		if ((regs->UECON[ep_idx] & USBC_UECON0_FIFOCON) == 0) {
+			return -EAGAIN;
+		}
+	}
+
+	ep_bank = usb_dc_sam_usbc_ep_curr_bank(ep_idx);
+
+	packet_len = MIN(data_len, dev_data.ep_data[ep_idx].mps);
+
+	if (data && packet_len > 0) {
+		memcpy(dev_desc[ep_bank].ep_pipe_addr, data, packet_len);
+		__DSB();
+	}
+	dev_desc[ep_bank].sizes = packet_len;
+
+	if (ep_idx == 0U) {
+		key = irq_lock();
+		/*
+		 * Control endpoint: clear the interrupt flag to send the
+		 * data, and re-enable the interrupts to trigger an interrupt
+		 * at the end of the transfer.
+		 */
+		regs->UESTACLR[0] = USBC_UESTA0CLR_TXINIC;
+		regs->UECONSET[0] = USBC_UECON0SET_TXINES;
+
+		if (packet_len == 0) {
+			/* To detect a protocol error, enable nak interrupt
+			 * on data OUT phase
+			 */
+			regs->UESTACLR[0] = USBC_UESTA0CLR_NAKOUTIC;
+			regs->UECONSET[0] = USBC_UECON0SET_NAKOUTES;
+		}
+		irq_unlock(key);
+	} else {
+		/*
+		 * Other endpoint types: clear the FIFO control flag to send
+		 * the data.
+		 */
+		regs->UECONCLR[ep_idx] = USBC_UECON0CLR_FIFOCONC;
+	}
+
+	if (ret_bytes) {
+		*ret_bytes = packet_len;
+	}
+
+	LOG_DBG("ep 0x%02x write %d bytes from %d to bank %d%s",
+		ep, packet_len, data_len, ep_bank % 2,
+		packet_len == 0 ? " (ZLP)" : "");
+
+	return 0;
+}
+
+int usb_dc_ep_read_ex(uint8_t ep, uint8_t *data, uint32_t max_data_len,
+		      uint32_t *read_bytes, bool wait)
+{
+	uint8_t ep_idx = EP_ADDR2IDX(ep);
+	uint8_t ep_bank;
+	uint32_t data_len;
+	uint32_t remaining;
+	uint32_t take;
+	int rc = 0;
+
+	if (ep_idx >= NUM_OF_EP_MAX) {
+		LOG_ERR("wrong endpoint index/address");
+		return -EINVAL;
+	}
+
+	if (!usb_dc_ep_is_enabled(ep_idx)) {
+		LOG_ERR("endpoint not enabled");
+		return -ENODEV;
+	}
+
+	if (EP_ADDR2DIR(ep) != USB_EP_DIR_OUT) {
+		LOG_ERR("wrong endpoint direction");
+		return -EINVAL;
+	}
+
+	if ((regs->UECON[ep_idx] & USBC_UECON0_STALLRQ) != 0) {
+		LOG_WRN("endpoint is stalled");
+		return -EBUSY;
+	}
+
+	ep_bank = usb_dc_sam_usbc_ep_curr_bank(ep_idx);
+	data_len = dev_desc[ep_bank].SIZES.byte_count;
+
+	/*
+	 * The code below emulates the Quark FIFO which the Zephyr USB
+	 * API is based on.  Reading with buf == NULL returns the
+	 * number of bytes available and starts the read.  The caller
+	 * then keeps calling until all bytes are consumed which
+	 * also marks the OUT buffer as freed.
+	 */
```

Comment:
please remove comment regarding Quark FIFO, current device driver API just expects this behavior.